### ISSUE TITLE
[Feature Request] frozen Cargo.lock for executable project

### ISSUE THREAD

**Issue by namse** [rust-lang/rust-analyzer#12219]
By [this document](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html), executable application check Cargo.lock into git.

Rust analyzer use `cargo check` command, but it updates Cargo.lock.

Sometime when I want to change git branch, it keep changing Cargo.lock, but I don't want to update it, so even I rollback Cargo.lock, but rust analyzer update it again!

I hope Rust analyzer support not to update Cargo.toml for some project if they put that into configuration.

**Comment by lnicola** [comment]
We run both `cargo check` and `cargo metadata`. You can override (or disable) the first to add `--locked`, but not the second.

**Comment by lnicola** [comment]
But I'm not sure this helps as much as it may seem. Sometimes people will just change `Cargo.toml` without compiling the code, and then you'll need to update the lockfile anyway.

**Comment by namse** [comment]
I think this problem is caused by rust analyzer updating Cargo.lock.

I think users should update Cargo.lock on their own. For example, the user updates using `cargo build` etc.

So I don't think it's a very good idea for a 3rd party rust analyzer to update their project's Cargo.lock. Because it is difficult to think that Cargo.lock will be automatically updated if the user turns on the rust analyzer.

The biggest problem is, I'm using the vscode workspace, and it's hard not to update the Cargo.lock for a specific project.

I tried to put --frozen or --locked in `rust-analyzer.checkOnSave.extraArgs` in a specific rust project, but vscode gave an error saying "This option can only be put in the root of the workspace". Ulala. I don't want to do this for all projects in the workspace, I just want to do this for a specific executable rust project!?

**Comment by namse** [comment]
Perhaps the more fundamental problem is that Cargo automatically installs the packages.

I often use npm in nodejs, where I explicitly install packages like `npm install`. Never let a 3rd party program automatically install the package.
Disadvantages also exist. They should always do `npm install` when doing a git pull or something to update local-installed packages. But I think it's more useful, at least because package-lock.json (npm's Cargo.lock) is not updated just because they saw the code in vscode.

I know it's pointless to talk about the downsides to Cargo as this is not a Cargo repository, but I'm still a fan of the rust analyzer, so I'd like to see if we can circumvent this issue.

**Comment by lnicola** [comment]
You can disable `cargo check`, disable proc macros and build scripts, and maybe enable our experimental diagnostics, which include type checking. That should alleviate any security concerns, but it won't be pleasant.

**Comment by namse** [comment]
How can I disable cargo check and use experimental diagnostics?

I have other idea. It can be a very complicated idea, but if Cargo.lock is included in git, how about adding --frozen or --locked as parameters when doing `cargo check`, etc.?

**Comment by lnicola** [comment]
`rust-analyzer.checkOnSave.enable`, `rust-analyzer.procMacro.enable`, `rust-analyzer.cargo.runBuildScripts` and `rust-analyzer.diagnostics.enableExperimental`. I also like to do:

```json
    "editor.semanticTokenColorCustomizations": {
        "rules": {
            "unresolvedReference": "#ff0000",
        },
    }
```

, which highlights unresolved references. Of course, you'll have a lot of those if you disable proc macros and build scripts.

> I have other idea. It can be a very complicated idea, but if Cargo.lock is included in git, how about adding --frozen or --locked as parameters when doing cargo check, etc.?

You just tried that and it appears to not apply to workspaces.

**Comment by namse** [comment]
I switch git branch like `git checkout -f && git checkout {branch}` to avoid Cargo.lock problem. it's so desperate...

**Comment by namse** [comment]
I know it's good to update `Cargo.lock` if `cargo check` updates it, but I mean, when I do something in git branch works, I don't want to update it during that time, I just want to fix my branches and commits. 
But Rust-analyzer doesn't know my mind and it keep update `Cargo.lock`. 

For example, I locked master branch pushing so I never update code in master branch. But need to pull remote master branch so I usually switch branch into master. So I pulled it and want to back to my dev branch but nah, hello new fresh Cargo.lock, it blocks me! no, no, not this time dude, I cannot push you from this master branch. but they don't listen to my voice, keeping update it even I remove it.

**Comment by flodiebold** [comment]
Having a master branch where the checked-in `Cargo.lock` doesn't match the `Cargo.toml` completely negates the benefits of checking in `Cargo.lock` in the first place. If you regularly are in this situation, you should address it in some way, and I don't think there is much utility in us adding special handling for it.

**Comment by namse** [comment]
I don't agree 100%. `Cargo.lock` locks the dependencies of executable so that they can always produce the same build. So it should not always update. It have to frozen in some cases, like finding bug, building the same dependencies with distributed executable.

This is not my own argument. That's what the cargo documentation says.

> If you build this package today, and then you send a copy to me, and I build this package tomorrow, something bad could happen. There could be more commits to regex in the meantime, and my build would include new commits while yours would not. Therefore, we would get different builds. This would be bad because we want reproducible builds.

https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html

**Comment by flodiebold** [comment]
The point is that `cargo check` and `cargo metadata` *only* update the file when the packages and versions in `Cargo.lock` cannot match the ones described in `Cargo.toml`, so for example when you add new dependencies or change the versions in `Cargo.toml`. Only `cargo update` will update the lockfile otherwise. I.e. if you run `cargo check` and commit the lockfile together with the `Cargo.toml`, it will never change again.

If you check in such a lockfile that has to be updated by `cargo check`, you *don't* get reproducible builds, because Cargo has to update it to actually build the project and might make different choices at different times. That's what I meant by "completely negates the benefits of checking in `Cargo.lock` in the first place".

**Comment by namse** [comment]
AH.............!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by namse** [comment]
I apologize for my ignorance. I didn't know about that, I assumed they update automatically my dependencies.......!

**Comment by namse** [comment]
To make panic when someone didn't update Cargo.lock after fixing Cargo.toml?
